### DISCUSSION TITLE
How to configure where to store assets using a single IO manager.

### DISCUSSION THREAD

**Discussion by JensvandeZande** [dagster-io/dagster#14852] [Q&A]
Hi all.

I have the following use case:

- I work locally.
- I want to create medallion like delta lake (bronze, silver, gold).
- I want to be able to specify where on my local file system to store each asset (delta table)
    - Example: store asset "TableA" at file location "my_project/data/silver/database/TableA/" as delta table
- I intend to use a custom IO-Manager i.e., DeltaIOManager
- Imagine for each layer (bronze, silver, gold) I have 20 different databases.

If I understood the [dagster tutorial](https://docs.dagster.io/tutorial/saving-your-data) correctly I would have to instantiate 60 different DeltaIOManagers(), i.e. 20 databases for each bronze, silver and gold layer. As you can imagine, this is not really feasible and a lot of work. Is there another method to accomplishing this? For example specifying the path at asset level?

Your help would be much appreciated!

KR Jens

**Comment by JensvandeZande** [comment]
@sryza I am pinging you because my question is not answered yet. Could you provide me help?

**Comment by sryza** [comment]
Hi @JensvandeZande - you could put the location in the metadata and access it from your IOManager:

```python
@asset(path="a/b/c", io_manager_key="delta_io_manager")
def asset1():
    ...

class DeltaIOManager():
    def handle_output(context):
        path = context.metadata["path"]
        ...
```

**Reply by JensvandeZande** [reply]
Thank you very much @sryza! How would you do the same using a run config?

**Comment by Beaglefoot** [comment]
You could also utilize asset keys and prefixes. For example:
```py
@asset(
    name="TableA",
    key_prefix=["silver", "database"],
    io_manager_key="delta_io_manager"
)
def asset1():
    ...

class DeltaIOManager():
    common_prefix = "my_project/data"

    def handle_output(context):
        path = os.path.join(common_prefix, *context.asset_key.path)
        ...

    def load_input(self, context: InputContext):
        path = os.path.join(common_prefix, *context.asset_key.path)
        ...
```
This way assets have a nice representation in Web UI and your naming conventions for local file system are centralized in one place.

**Reply by JensvandeZande** [reply]
@Beaglefoot many thanks for your input! The only thing I try to solve is that in my opinion the location where the data is stored is part of configuration. So I would like to be able to change the location of the data without changing the code. What do you think?

**Reply by Beaglefoot** [reply]
Definitely possible. Here's what you can do:
```py
class MyRunConfig(Config): ...

class ExternalDeltaIOManager():
    def __init__(self, config: MyRunConfig):
        self.config = config

    def handle_output(context: OutputContext): ...

    def load_input(self, context: InputContext): ...

class DeltaIOManager(ConfigurableIOManagerFactory):
    def create_io_manager(self, context: InitResourceContext) -> ExternalDeltaIOManager:
        config = cast(MyRunConfig, context.dagster_run.run_config)

        ...

        return ExternalDeltaIOManager(config)
```
Although things get complicated when you want to make your io manager compatible both with Pythonic and Legacy run configs.

**Reply by JensvandeZande** [reply]
I am sorry @Beaglefoot, I fail to understand how it works. How would dagster know to use the right configuration fields from my config.yaml?

Let me show you my code and perhaps you could help me?

So what I am trying to do, is to supply configuration to my assets using the config argument. I would then like to access that config from within the IO manager using the InputContext and OutputContext. However the `context.config` object is of type NoneType. I have the feeling I do not fully understand how configuration and contexts work. The documentation on these topics is very limited. Hopefully you can help?

```python 
import datadotworld as dw
from dagster import ConfigurableIOManager, Definitions, InputContext, MetadataValue, Output, OutputContext, asset
from wine.config import Conf
from dagster import Config
from typing import List, Union
import pandas as pd
from deltalake.writer import write_deltalake
from deltalake import DeltaTable


class AssetConf(Config):
    path: str

class Conf(Config):
    asset_conf: AssetConf


@asset(io_manager_key="pandas_delta_io")
def data_world(config:Conf):
    path = "markpowell/global-wine-points"
    data = "wines"
    dw_wine = dw.query(path, f"SELECT * FROM {data}").dataframe
    return Output(
        value=dw_wine,
        metadata={
            "num_records": len(dw_wine),
            "preview": MetadataValue.md(dw_wine.head().to_markdown()),
            "location": config.asset_conf.path
        },
    )
    
class PandasDeltaIOManager(ConfigurableIOManager):
    path_prefix: List[str] = []

    def _get_path(self, context: Union[OutputContext, InputContext]) -> str:
        return "/".join(self.path_prefix + [context.config["path"]] + [context.asset_key.path[-1]])

    def handle_output(self, context: OutputContext, obj:pd.DataFrame):
        context.add_output_metadata({"file_path": self._get_path(context)})
        write_deltalake(self._get_path(context=context), data=obj, mode="overwrite")

    def load_input(self, context: InputContext):
        return DeltaTable(self._get_path(context)).to_pandas()


defs = Definitions(
    assets=data_world,
    resources= {
        "pandas_delta_io": PandasDeltaIOManager(path_prefix=["bronze"])
    }
)
```
